Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,613 @@
+    /// a request has its `EventLoopPreference` property set to `.indifferent`.
+    /// Having a default `EventLoop` shared by the *channel* and the *delegate* avoids
+    /// loss of performance due to `EventLoop` hoping
+    func associatedEventLoop(for key: Key) -> EventLoop? {
+        return self[key]?.eventLoop
+    }
+
+    /// Determines what action should be taken regarding the `ConnectionProvider`
+    ///
+    /// This method ensures there is no race condition if called concurrently from multiple threads
+    private func getAction(for key: Key, on eventLoop: EventLoop) -> ProviderGetAction {
+        return self.connectionProvidersLock.withLock {
+            if let provider = self._connectionProviders[key] {
+                return .useExisting(provider)
+            } else {
+                let promise = eventLoop.makePromise(of: ConnectionProvider.self)
+                self._connectionProviders[key] = .future(promise.futureResult)
+                promise.futureResult.whenComplete { result in
+                    switch result {
+                    case .success(let provider):
+                        self[key] = provider
+                    case .failure:
+                        self[key] = nil
+                    }
+                }
+                return .make(promise)
+            }
+        }
+    }
+
+    /// This method asks the pool for a connection usable by the specified `request`, respecting the specified options.
+    ///
+    /// - parameter request: The request that needs a `Connection`
+    /// - parameter preference: The `EventLoopPreference` the connection pool will respect to lease a new connection
+    /// - parameter deadline: The connection timeout
+    /// - Returns: A connection  corresponding to the specified parameters
+    ///
+    /// When the pool is asked for a new connection, it creates a `Key` from the url associated to the `request`. This key
+    /// is used to determine if there already exists an associated `ConnectionProvider` in `connectionProviders`
+    /// if it does, the connection provider then takes care of leasing a new connection. If a connection provider doesn't exist, it is created.
+    func getConnection(for request: HTTPClient.Request, preference: HTTPClient.EventLoopPreference, on eventLoop: EventLoop, deadline: NIODeadline?) -> EventLoopFuture<Connection> {
+        let key = Key(request)
+        switch self.getAction(for: key, on: eventLoop) {
+        case .useExisting(let provider):
+            return provider.getConnection(preference: preference)
+
+        case .make(completing: let providerPromise):
+            // If no connection provider exists for the given key, create a new one
+            let selectedEventLoop = providerPromise.futureResult.eventLoop
+            var bootstrap = ClientBootstrap.makeHTTPClientBootstrapBase(group: selectedEventLoop, host: request.host, port: request.port, configuration: self.configuration)
```

Comment:
this is duplicated in `HTTP1ConnectionProvider.makeConnection`. Can we create an empty `HTTP1ConnectionProvider` and then call `getConnection` to create a new connection? That way this code will not be duplicated.

Reply:
@artemredkin Thank you, I'll try this approach! 🙂

Reply:
@artemredkin Started modifying this in https://github.com/swift-server/async-http-client/pull/105/commits/5d3b0cc34d31e02dad6b9b2c7a9bc7de48b2abbf, I will then completely remove `ConnectionProvider`.

Reply:
`ConnectionProvider` is now gone in https://github.com/swift-server/async-http-client/pull/105/commits/bc62a077fa3eadc5a69b981055de5a0e59822b29

Thank you very much @artemredkin for suggesting getting rid of the `.future` case